File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0;
        int r = heights.size() - 1;

        int maxWater = 0;

        while (l < r) {
            maxWater = std::max(maxWater, std::min(heights[l], heights[r]) * (r - l));
            if (heights[l] < heights[r]) l++;
            else r--;
        }

        return maxWater;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0

        l = 0
        r = 0
        last_seen = {}
        longest_sub = 1
        
        while r < len(s):
            if last_seen.get(s[r], -1) >= l:
                l += 1
            else:
                last_seen[s[r]] = r
                longest_sub = max(longest_sub, r - l + 1)
                r += 1
        
        return longest_sub